internal/controller/k0smotron.io/k0smotroncluster_etcd.go

```diff
@@ -0,0 +1,357 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package k0smotronio
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	km "github.com/k0sproject/k0smotron/api/k0smotron.io/v1beta1"
+
+	apps "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/cluster-api/util/secret"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *ClusterReconciler) reconcileEtcd(ctx context.Context, kmc *km.Cluster) error {
+	if kmc.Spec.KineDataSourceURL != "" || kmc.Spec.KineDataSourceSecretName != "" {
+		return nil
+	}
+
+	if err := r.reconcileEtcdSvc(ctx, kmc); err != nil {
+		return fmt.Errorf("error reconciling etcd service: %w", err)
+	}
+	if err := r.reconcileEtcdStatefulSet(ctx, kmc); err != nil {
+		return fmt.Errorf("error reconciling etcd statefulset: %w", err)
+	}
+
+	return nil
+}
+
+func (r *ClusterReconciler) reconcileEtcdSvc(ctx context.Context, kmc *km.Cluster) error {
+	labels := labelsForEtcdCluster(kmc)
+
+	svc := v1.Service{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Service",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        kmc.GetEtcdServiceName(),
+			Namespace:   kmc.Namespace,
+			Labels:      labels,
+			Annotations: annotationsForCluster(kmc),
+		},
+		Spec: v1.ServiceSpec{
+			Type:                     v1.ServiceTypeClusterIP,
+			ClusterIP:                v1.ClusterIPNone,
+			Selector:                 labels,
+			PublishNotReadyAddresses: true,
+			Ports: []v1.ServicePort{
+				{
+					Name:       "client",
+					Port:       2379,
+					TargetPort: intstr.FromInt32(2379),
+				},
+				{
+					Name:       "peer",
+					Port:       2380,
+					TargetPort: intstr.FromInt32(2380),
+				},
+			},
+		},
+	}
+
+	_ = ctrl.SetControllerReference(kmc, &svc, r.Scheme)
+
+	return r.Client.Patch(ctx, &svc, client.Apply, patchOpts...)
+}
+
+func (r *ClusterReconciler) reconcileEtcdStatefulSet(ctx context.Context, kmc *km.Cluster) error {
```

Comment:
The logic is bit hard to follow, maybe few comments would be in order.

Also if we'd separate the "calculation" into separate func we could also have proper unit tests for it. Tests also makes it easier to understand, IMO.

Reply:
Added some comments and tests